Not compiling with gcc >= 9

Hi,

gcc 9 seems to introduce warnings which make the build fail.
In particular, `-Wdeprecated-copy`, which forbids to use `Foo operator=` is `Foo& operator=` is defined generates a couple of errors in `libdevcore/FixedHash.h` because

```cpp
SecureFixedHash<T>& operator=(SecureFixedHash<T> const& _c)
```

is defined but `SecureFixedHash<T> operator=` is used here and there.

Another warning newly triggered is `stringop-truncation` because of `strncpy` at in `libweb3jsonrpc/UnixSocketServer.cpp:UnixDomainSocketServer::StartListening`.

I am not sure what is the best way to go so for now, I simply disabled the warnings, globally for `deprecated-copy`:

```cmake
# cmake/EthCompilerSettings.cmake
add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-deprecated-copy>)
```

and locally for `stringop-truncation`:

```cpp
// libweb3jsonrpc/UnixSocketServer.cpp

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wstringop-truncation"
strncpy(m_address.sun_path, m_path.c_str(), c_socketPathMaxLength);
#pragma GCC diagnostic pop
```

More information about the new warnings:
https://gcc.gnu.org/gcc-9/changes.html

Once you pass the `FixedHash<T>` error, there is second one:

```
/home/mwo2/aleth/libweb3jsonrpc/UnixSocketServer.cpp: In member function ‘virtual bool dev::UnixDomainSocketServer::StartListening()’:
/home/mwo2/aleth/libweb3jsonrpc/UnixSocketServer.cpp:85:10: error: ‘char* strncpy(char*, const char*, size_t)’ specified bound 108 equals destination size [-Werror=stringop-truncation]
   85 |   strncpy(m_address.sun_path, m_path.c_str(), c_socketPathMaxLength);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [libweb3jsonrpc/CMakeFiles/web3jsonrpc.dir/build.make:219: libweb3jsonrpc/CMakeFiles/web3jsonrpc.dir/UnixSocketServer.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:989: libweb3jsonrpc/CMakeFiles/web3jsonrpc.dir/all] Error 2
```

--- a/libdevcore/FixedHash.h
+++ b/libdevcore/FixedHash.h
@@ -219,6 +219,7 @@ public:
     using ConstructFromStringType = typename FixedHash<T>::ConstructFromStringType;
     using ConstructFromPointerType = typename FixedHash<T>::ConstructFromPointerType;
     SecureFixedHash() = default;
+    SecureFixedHash(SecureFixedHash const&) = default;
     explicit SecureFixedHash(bytes const& _b, ConstructFromHashType _t = FixedHash<T>::FailIfDifferent): FixedHash<T>(_b, _t) {}
     explicit SecureFixedHash(bytesConstRef _b, ConstructFromHashType _t = FixedHash<T>::FailIfDifferent): FixedHash<T>(_b, _t) {}
     explicit SecureFixedHash(bytesSec const& _b, ConstructFromHashType _t = FixedHash<T>::FailIfDifferent): FixedHash<T>(_b.ref(), _t) {}

--- a/libdevcore/FixedHash.h
+++ b/libdevcore/FixedHash.h
@@ -29,304 +29,304 @@ template <unsigned N>
 class FixedHash
 {
 public:
-	/// The corresponding arithmetic type.
-	using Arith = boost::multiprecision::number<boost::multiprecision::cpp_int_backend<N * 8, N * 8, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>>;
+    /// The corresponding arithmetic type.
+    using Arith = boost::multiprecision::number<boost::multiprecision::cpp_int_backend<N * 8, N * 8, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>>;
 
-	/// The size of the container.
-	enum { size = N };
+    /// The size of the container.
+    enum { size = N };
 
-	/// A dummy flag to avoid accidental construction from pointer.
-	enum ConstructFromPointerType { ConstructFromPointer };
+    /// A dummy flag to avoid accidental construction from pointer.
+    enum ConstructFromPointerType { ConstructFromPointer };
 
-	/// Method to convert from a string.
-	enum ConstructFromStringType { FromHex, FromBinary };
+    /// Method to convert from a string.
+    enum ConstructFromStringType { FromHex, FromBinary };
 
-	/// Method to convert from a string.
-	enum ConstructFromHashType { AlignLeft, AlignRight, FailIfDifferent };
+    /// Method to convert from a string.
+    enum ConstructFromHashType { AlignLeft, AlignRight, FailIfDifferent };
 
-	/// Construct an empty hash.
-	FixedHash() { m_data.fill(0); }
+    /// Construct an empty hash.
+    FixedHash() { m_data.fill(0); }
 
-	/// Construct from another hash, filling with zeroes or cropping as necessary.
-	template <unsigned M> explicit FixedHash(FixedHash<M> const& _h, ConstructFromHashType _t = AlignLeft) { m_data.fill(0); unsigned c = std::min(M, N); for (unsigned i = 0; i < c; ++i) m_data[_t == AlignRight ? N - 1 - i : i] = _h[_t == AlignRight ? M - 1 - i : i]; }
+    /// Construct from another hash, filling with zeroes or cropping as necessary.
+    template <unsigned M> explicit FixedHash(FixedHash<M> const& _h, ConstructFromHashType _t = AlignLeft) { m_data.fill(0); unsigned c = std::min(M, N); for (unsigned i = 0; i < c; ++i) m_data[_t == AlignRight ? N - 1 - i : i] = _h[_t == AlignRight ? M - 1 - i : i]; }
 
-	/// Convert from the corresponding arithmetic type.
-	FixedHash(Arith const& _arith) { toBigEndian(_arith, m_data); }
+    /// Convert from the corresponding arithmetic type.
+    FixedHash(Arith const& _arith) { toBigEndian(_arith, m_data); }
 
-	/// Convert from unsigned
-	explicit FixedHash(unsigned _u) { toBigEndian(_u, m_data); }
+    /// Convert from unsigned
+    explicit FixedHash(unsigned _u) { toBigEndian(_u, m_data); }
 
-	/// Explicitly construct, copying from a byte array.
-	explicit FixedHash(bytes const& _b, ConstructFromHashType _t = FailIfDifferent) { if (_b.size() == N) memcpy(m_data.data(), _b.data(), std::min<unsigned>(_b.size(), N)); else { m_data.fill(0); if (_t != FailIfDifferent) { auto c = std::min<unsigned>(_b.size(), N); for (unsigned i = 0; i < c; ++i) m_data[_t == AlignRight ? N - 1 - i : i] = _b[_t == AlignRight ? _b.size() - 1 - i : i]; } } }
+    /// Explicitly construct, copying from a byte array.
+    explicit FixedHash(bytes const& _b, ConstructFromHashType _t = FailIfDifferent) { if (_b.size() == N) memcpy(m_data.data(), _b.data(), std::min<unsigned>(_b.size(), N)); else { m_data.fill(0); if (_t != FailIfDifferent) { auto c = std::min<unsigned>(_b.size(), N); for (unsigned i = 0; i < c; ++i) m_data[_t == AlignRight ? N - 1 - i : i] = _b[_t == AlignRight ? _b.size() - 1 - i : i]; } } }
 
-	/// Explicitly construct, copying from a byte array.
-	explicit FixedHash(bytesConstRef _b, ConstructFromHashType _t = FailIfDifferent) { if (_b.size() == N) memcpy(m_data.data(), _b.data(), std::min<unsigned>(_b.size(), N)); else { m_data.fill(0); if (_t != FailIfDifferent) { auto c = std::min<unsigned>(_b.size(), N); for (unsigned i = 0; i < c; ++i) m_data[_t == AlignRight ? N - 1 - i : i] = _b[_t == AlignRight ? _b.size() - 1 - i : i]; } } }
+    /// Explicitly construct, copying from a byte array.
+    explicit FixedHash(bytesConstRef _b, ConstructFromHashType _t = FailIfDifferent) { if (_b.size() == N) memcpy(m_data.data(), _b.data(), std::min<unsigned>(_b.size(), N)); else { m_data.fill(0); if (_t != FailIfDifferent) { auto c = std::min<unsigned>(_b.size(), N); for (unsigned i = 0; i < c; ++i) m_data[_t == AlignRight ? N - 1 - i : i] = _b[_t == AlignRight ? _b.size() - 1 - i : i]; } } }
 
-	/// Explicitly construct, copying from a bytes in memory with given pointer.
-	explicit FixedHash(byte const* _bs, ConstructFromPointerType) { memcpy(m_data.data(), _bs, N); }
+    /// Explicitly construct, copying from a bytes in memory with given pointer.
+    explicit FixedHash(byte const* _bs, ConstructFromPointerType) { memcpy(m_data.data(), _bs, N); }
 
-	/// Explicitly construct, copying from a  string.
-	explicit FixedHash(std::string const& _s, ConstructFromStringType _t = FromHex, ConstructFromHashType _ht = FailIfDifferent): FixedHash(_t == FromHex ? fromHex(_s, WhenError::Throw) : dev::asBytes(_s), _ht) {}
+    /// Explicitly construct, copying from a  string.
+    explicit FixedHash(std::string const& _s, ConstructFromStringType _t = FromHex, ConstructFromHashType _ht = FailIfDifferent): FixedHash(_t == FromHex ? fromHex(_s, WhenError::Throw) : dev::asBytes(_s), _ht) {}
 
-	/// Convert to arithmetic type.
-	operator Arith() const { return fromBigEndian<Arith>(m_data); }
+    /// Convert to arithmetic type.
+    operator Arith() const { return fromBigEndian<Arith>(m_data); }
 
-	/// @returns true iff this is the empty hash.
-	explicit operator bool() const { return std::any_of(m_data.begin(), m_data.end(), [](byte _b) { return _b != 0; }); }
+    /// @returns true iff this is the empty hash.
+    explicit operator bool() const { return std::any_of(m_data.begin(), m_data.end(), [](byte _b) { return _b != 0; }); }
 
-	// The obvious comparison operators.
-	bool operator==(FixedHash const& _c) const { return m_data == _c.m_data; }
-	bool operator!=(FixedHash const& _c) const { return m_data != _c.m_data; }
-	bool operator<(FixedHash const& _c) const { for (unsigned i = 0; i < N; ++i) if (m_data[i] < _c.m_data[i]) return true; else if (m_data[i] > _c.m_data[i]) return false; return false; }
-	bool operator>=(FixedHash const& _c) const { return !operator<(_c); }
-	bool operator<=(FixedHash const& _c) const { return operator==(_c) || operator<(_c); }
-	bool operator>(FixedHash const& _c) const { return !operator<=(_c); }
+    // The obvious comparison operators.
+    bool operator==(FixedHash const& _c) const { return m_data == _c.m_data; }
+    bool operator!=(FixedHash const& _c) const { return m_data != _c.m_data; }
+    bool operator<(FixedHash const& _c) const { for (unsigned i = 0; i < N; ++i) if (m_data[i] < _c.m_data[i]) return true; else if (m_data[i] > _c.m_data[i]) return false; return false; }
+    bool operator>=(FixedHash const& _c) const { return !operator<(_c); }
+    bool operator<=(FixedHash const& _c) const { return operator==(_c) || operator<(_c); }
+    bool operator>(FixedHash const& _c) const { return !operator<=(_c); }
 
-	// The obvious binary operators.
-	FixedHash& operator^=(FixedHash const& _c) { for (unsigned i = 0; i < N; ++i) m_data[i] ^= _c.m_data[i]; return *this; }
-	FixedHash operator^(FixedHash const& _c) const { return FixedHash(*this) ^= _c; }
-	FixedHash& operator|=(FixedHash const& _c) { for (unsigned i = 0; i < N; ++i) m_data[i] |= _c.m_data[i]; return *this; }
-	FixedHash operator|(FixedHash const& _c) const { return FixedHash(*this) |= _c; }
-	FixedHash& operator&=(FixedHash const& _c) { for (unsigned i = 0; i < N; ++i) m_data[i] &= _c.m_data[i]; return *this; }
-	FixedHash operator&(FixedHash const& _c) const { return FixedHash(*this) &= _c; }
-	FixedHash operator~() const { FixedHash ret; for (unsigned i = 0; i < N; ++i) ret[i] = ~m_data[i]; return ret; }
+    // The obvious binary operators.
+    FixedHash& operator^=(FixedHash const& _c) { for (unsigned i = 0; i < N; ++i) m_data[i] ^= _c.m_data[i]; return *this; }
+    FixedHash operator^(FixedHash const& _c) const { return FixedHash(*this) ^= _c; }
+    FixedHash& operator|=(FixedHash const& _c) { for (unsigned i = 0; i < N; ++i) m_data[i] |= _c.m_data[i]; return *this; }
+    FixedHash operator|(FixedHash const& _c) const { return FixedHash(*this) |= _c; }
+    FixedHash& operator&=(FixedHash const& _c) { for (unsigned i = 0; i < N; ++i) m_data[i] &= _c.m_data[i]; return *this; }
+    FixedHash operator&(FixedHash const& _c) const { return FixedHash(*this) &= _c; }
+    FixedHash operator~() const { FixedHash ret; for (unsigned i = 0; i < N; ++i) ret[i] = ~m_data[i]; return ret; }
 
-	// Big-endian increment.
-	FixedHash& operator++() { for (unsigned i = size; i > 0 && !++m_data[--i]; ) {} return *this; }
+    // Big-endian increment.
+    FixedHash& operator++() { for (unsigned i = size; i > 0 && !++m_data[--i]; ) {} return *this; }
 
-	/// @returns true if all one-bits in @a _c are set in this object.
-	bool contains(FixedHash const& _c) const { return (*this & _c) == _c; }
+    /// @returns true if all one-bits in @a _c are set in this object.
+    bool contains(FixedHash const& _c) const { return (*this & _c) == _c; }
 
-	/// @returns a particular byte from the hash.
-	byte& operator[](unsigned _i) { return m_data[_i]; }
-	/// @returns a particular byte from the hash.
-	byte operator[](unsigned _i) const { return m_data[_i]; }
+    /// @returns a particular byte from the hash.
+    byte& operator[](unsigned _i) { return m_data[_i]; }
+    /// @returns a particular byte from the hash.
+    byte operator[](unsigned _i) const { return m_data[_i]; }
 
-	/// @returns an abridged version of the hash as a user-readable hex string.
-	std::string abridged() const { return toHex(ref().cropped(0, 4)) + "\342\200\246"; }
+    /// @returns an abridged version of the hash as a user-readable hex string.
+    std::string abridged() const { return toHex(ref().cropped(0, 4)) + "\342\200\246"; }
 
-	/// @returns a version of the hash as a user-readable hex string that leaves out the middle part.
-	std::string abridgedMiddle() const { return toHex(ref().cropped(0, 4)) + "\342\200\246" + toHex(ref().cropped(N - 4)); }
+    /// @returns a version of the hash as a user-readable hex string that leaves out the middle part.
+    std::string abridgedMiddle() const { return toHex(ref().cropped(0, 4)) + "\342\200\246" + toHex(ref().cropped(N - 4)); }
 
-	/// @returns the hash as a user-readable hex string.
-	std::string hex() const { return toHex(ref()); }
+    /// @returns the hash as a user-readable hex string.
+    std::string hex() const { return toHex(ref()); }
 
-	/// @returns a mutable byte vector_ref to the object's data.
-	bytesRef ref() { return bytesRef(m_data.data(), N); }
+    /// @returns a mutable byte vector_ref to the object's data.
+    bytesRef ref() { return bytesRef(m_data.data(), N); }
 
-	/// @returns a constant byte vector_ref to the object's data.
-	bytesConstRef ref() const { return bytesConstRef(m_data.data(), N); }
+    /// @returns a constant byte vector_ref to the object's data.
+    bytesConstRef ref() const { return bytesConstRef(m_data.data(), N); }
 
-	/// @returns a mutable byte pointer to the object's data.
-	byte* data() { return m_data.data(); }
+    /// @returns a mutable byte pointer to the object's data.
+    byte* data() { return m_data.data(); }
 
-	/// @returns a constant byte pointer to the object's data.
-	byte const* data() const { return m_data.data(); }
+    /// @returns a constant byte pointer to the object's data.
+    byte const* data() const { return m_data.data(); }
 
-	/// @returns begin iterator.
-	auto begin() const -> typename std::array<byte, N>::const_iterator { return m_data.begin(); }
+    /// @returns begin iterator.
+    auto begin() const -> typename std::array<byte, N>::const_iterator { return m_data.begin(); }
 
-	/// @returns end iterator.
-	auto end() const -> typename std::array<byte, N>::const_iterator { return m_data.end(); }
+    /// @returns end iterator.
+    auto end() const -> typename std::array<byte, N>::const_iterator { return m_data.end(); }
 
-	/// @returns a copy of the object's data as a byte vector.
-	bytes asBytes() const { return bytes(data(), data() + N); }
+    /// @returns a copy of the object's data as a byte vector.
+    bytes asBytes() const { return bytes(data(), data() + N); }
 
-	/// @returns a mutable reference to the object's data as an STL array.
-	std::array<byte, N>& asArray() { return m_data; }
+    /// @returns a mutable reference to the object's data as an STL array.
+    std::array<byte, N>& asArray() { return m_data; }
 
-	/// @returns a constant reference to the object's data as an STL array.
-	std::array<byte, N> const& asArray() const { return m_data; }
+    /// @returns a constant reference to the object's data as an STL array.
+    std::array<byte, N> const& asArray() const { return m_data; }
 
-	/// Populate with random data.
-	template <class Engine>
-	void randomize(Engine& _eng)
-	{
-		for (auto& i: m_data)
-			i = (uint8_t)std::uniform_int_distribution<uint16_t>(0, 255)(_eng);
-	}
+    /// Populate with random data.
+    template <class Engine>
+    void randomize(Engine& _eng)
+    {
+        for (auto& i: m_data)
+            i = (uint8_t)std::uniform_int_distribution<uint16_t>(0, 255)(_eng);
+    }
 
-	/// @returns a random valued object.
-	static FixedHash random() { FixedHash ret; ret.randomize(s_fixedHashEngine); return ret; }
+    /// @returns a random valued object.
+    static FixedHash random() { FixedHash ret; ret.randomize(s_fixedHashEngine); return ret; }
 
-	struct hash
-	{
-		/// Make a hash of the object's data.
-		size_t operator()(FixedHash const& _value) const { return boost::hash_range(_value.m_data.cbegin(), _value.m_data.cend()); }
-	};
+    struct hash
+    {
+        /// Make a hash of the object's data.
+        size_t operator()(FixedHash const& _value) const { return boost::hash_range(_value.m_data.cbegin(), _value.m_data.cend()); }
+    };
 
-	template <unsigned P, unsigned M> inline FixedHash& shiftBloom(FixedHash<M> const& _h)
-	{
-		return (*this |= _h.template bloomPart<P, N>());
-	}
+    template <unsigned P, unsigned M> inline FixedHash& shiftBloom(FixedHash<M> const& _h)
+    {
+        return (*this |= _h.template bloomPart<P, N>());
+    }
 
-	template <unsigned P, unsigned M> inline bool containsBloom(FixedHash<M> const& _h)
-	{
-		return contains(_h.template bloomPart<P, N>());
-	}
+    template <unsigned P, unsigned M> inline bool containsBloom(FixedHash<M> const& _h)
+    {
+        return contains(_h.template bloomPart<P, N>());
+    }
 
-	template <unsigned P, unsigned M> inline FixedHash<M> bloomPart() const
-	{
-		unsigned const c_bloomBits = M * 8;
-		unsigned const c_mask = c_bloomBits - 1;
-		unsigned const c_bloomBytes = (StaticLog2<c_bloomBits>::result + 7) / 8;
+    template <unsigned P, unsigned M> inline FixedHash<M> bloomPart() const
+    {
+        unsigned const c_bloomBits = M * 8;
+        unsigned const c_mask = c_bloomBits - 1;
+        unsigned const c_bloomBytes = (StaticLog2<c_bloomBits>::result + 7) / 8;
 
-		static_assert((M & (M - 1)) == 0, "M must be power-of-two");
-		static_assert(P * c_bloomBytes <= N, "out of range");
+        static_assert((M & (M - 1)) == 0, "M must be power-of-two");
+        static_assert(P * c_bloomBytes <= N, "out of range");
 
-		FixedHash<M> ret;
-		byte const* p = data();
-		for (unsigned i = 0; i < P; ++i)
-		{
-			unsigned index = 0;
-			for (unsigned j = 0; j < c_bloomBytes; ++j, ++p)
-				index = (index << 8) | *p;
-			index &= c_mask;
-			ret[M - 1 - index / 8] |= (1 << (index % 8));
-		}
-		return ret;
-	}
+        FixedHash<M> ret;
+        byte const* p = data();
+        for (unsigned i = 0; i < P; ++i)
+        {
+            unsigned index = 0;
+            for (unsigned j = 0; j < c_bloomBytes; ++j, ++p)
+                index = (index << 8) | *p;
+            index &= c_mask;
+            ret[M - 1 - index / 8] |= (1 << (index % 8));
+        }
+        return ret;
+    }
 
-	/// Returns the index of the first bit set to one, or size() * 8 if no bits are set.
-	inline unsigned firstBitSet() const
-	{
-		unsigned ret = 0;
-		for (auto d: m_data)
-			if (d)
-			{
-				for (;; ++ret, d <<= 1)
-				{
-					if (d & 0x80)
-						return ret;
-				}
-			}
-			else
-				ret += 8;
-		return ret;
-	}
+    /// Returns the index of the first bit set to one, or size() * 8 if no bits are set.
+    inline unsigned firstBitSet() const
+    {
+        unsigned ret = 0;
+        for (auto d: m_data)
+            if (d)
+            {
+                for (;; ++ret, d <<= 1)
+                {
+                    if (d & 0x80)
+                        return ret;
+                }
+            }
+            else
+                ret += 8;
+        return ret;
+    }
 
-	void clear() { m_data.fill(0); }
+    void clear() { m_data.fill(0); }
 
 private:
-	std::array<byte, N> m_data;		///< The binary data.
+    std::array<byte, N> m_data;        ///< The binary data.
 };
 
 template <unsigned T>
 class SecureFixedHash: private FixedHash<T>
 {
 public:
-	using ConstructFromHashType = typename FixedHash<T>::ConstructFromHashType;
-	using ConstructFromStringType = typename FixedHash<T>::ConstructFromStringType;
-	using ConstructFromPointerType = typename FixedHash<T>::ConstructFromPointerType;
-	SecureFixedHash() = default;
-	explicit SecureFixedHash(bytes const& _b, ConstructFromHashType _t = FixedHash<T>::FailIfDifferent): FixedHash<T>(_b, _t) {}
-	explicit SecureFixedHash(bytesConstRef _b, ConstructFromHashType _t = FixedHash<T>::FailIfDifferent): FixedHash<T>(_b, _t) {}
-	explicit SecureFixedHash(bytesSec const& _b, ConstructFromHashType _t = FixedHash<T>::FailIfDifferent): FixedHash<T>(_b.ref(), _t) {}
-	template <unsigned M> explicit SecureFixedHash(FixedHash<M> const& _h, ConstructFromHashType _t = FixedHash<T>::AlignLeft): FixedHash<T>(_h, _t) {}
-	template <unsigned M> explicit SecureFixedHash(SecureFixedHash<M> const& _h, ConstructFromHashType _t = FixedHash<T>::AlignLeft): FixedHash<T>(_h.makeInsecure(), _t) {}
-	explicit SecureFixedHash(std::string const& _s, ConstructFromStringType _t = FixedHash<T>::FromHex, ConstructFromHashType _ht = FixedHash<T>::FailIfDifferent): FixedHash<T>(_s, _t, _ht) {}
-	explicit SecureFixedHash(byte const* _d, ConstructFromPointerType _t): FixedHash<T>(_d, _t) {}
-	~SecureFixedHash() { ref().cleanse(); }
+    using ConstructFromHashType = typename FixedHash<T>::ConstructFromHashType;
+    using ConstructFromStringType = typename FixedHash<T>::ConstructFromStringType;
+    using ConstructFromPointerType = typename FixedHash<T>::ConstructFromPointerType;
+    SecureFixedHash() = default;
+    explicit SecureFixedHash(bytes const& _b, ConstructFromHashType _t = FixedHash<T>::FailIfDifferent): FixedHash<T>(_b, _t) {}
+    explicit SecureFixedHash(bytesConstRef _b, ConstructFromHashType _t = FixedHash<T>::FailIfDifferent): FixedHash<T>(_b, _t) {}
+    explicit SecureFixedHash(bytesSec const& _b, ConstructFromHashType _t = FixedHash<T>::FailIfDifferent): FixedHash<T>(_b.ref(), _t) {}
+    template <unsigned M> explicit SecureFixedHash(FixedHash<M> const& _h, ConstructFromHashType _t = FixedHash<T>::AlignLeft): FixedHash<T>(_h, _t) {}
+    template <unsigned M> explicit SecureFixedHash(SecureFixedHash<M> const& _h, ConstructFromHashType _t = FixedHash<T>::AlignLeft): FixedHash<T>(_h.makeInsecure(), _t) {}
+    explicit SecureFixedHash(std::string const& _s, ConstructFromStringType _t = FixedHash<T>::FromHex, ConstructFromHashType _ht = FixedHash<T>::FailIfDifferent): FixedHash<T>(_s, _t, _ht) {}
+    explicit SecureFixedHash(byte const* _d, ConstructFromPointerType _t): FixedHash<T>(_d, _t) {}
+    ~SecureFixedHash() { ref().cleanse(); }
 
-	SecureFixedHash<T>& operator=(SecureFixedHash<T> const& _c)
-	{
-		if (&_c == this)
-			return *this;
-		ref().cleanse();
-		FixedHash<T>::operator=(static_cast<FixedHash<T> const&>(_c));
-		return *this;
-	}
+    SecureFixedHash<T>& operator=(SecureFixedHash<T> const& _c)
+    {
+        if (&_c == this)
+            return *this;
+        ref().cleanse();
+        FixedHash<T>::operator=(static_cast<FixedHash<T> const&>(_c));
+        return *this;
+    }
 
-	using FixedHash<T>::size;
+    using FixedHash<T>::size;
 
-	bytesSec asBytesSec() const { return bytesSec(ref()); }
+    bytesSec asBytesSec() const { return bytesSec(ref()); }
 
-	FixedHash<T> const& makeInsecure() const { return static_cast<FixedHash<T> const&>(*this); }
-	FixedHash<T>& writable() { clear(); return static_cast<FixedHash<T>&>(*this); }
+    FixedHash<T> const& makeInsecure() const { return static_cast<FixedHash<T> const&>(*this); }
+    FixedHash<T>& writable() { clear(); return static_cast<FixedHash<T>&>(*this); }
 
-	using FixedHash<T>::operator bool;
+    using FixedHash<T>::operator bool;
 
-	// The obvious comparison operators.
-	bool operator==(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator==(static_cast<FixedHash<T> const&>(_c)); }
-	bool operator!=(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator!=(static_cast<FixedHash<T> const&>(_c)); }
-	bool operator<(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator<(static_cast<FixedHash<T> const&>(_c)); }
-	bool operator>=(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator>=(static_cast<FixedHash<T> const&>(_c)); }
-	bool operator<=(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator<=(static_cast<FixedHash<T> const&>(_c)); }
-	bool operator>(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator>(static_cast<FixedHash<T> const&>(_c)); }
+    // The obvious comparison operators.
+    bool operator==(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator==(static_cast<FixedHash<T> const&>(_c)); }
+    bool operator!=(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator!=(static_cast<FixedHash<T> const&>(_c)); }
+    bool operator<(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator<(static_cast<FixedHash<T> const&>(_c)); }
+    bool operator>=(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator>=(static_cast<FixedHash<T> const&>(_c)); }
+    bool operator<=(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator<=(static_cast<FixedHash<T> const&>(_c)); }
+    bool operator>(SecureFixedHash const& _c) const { return static_cast<FixedHash<T> const&>(*this).operator>(static_cast<FixedHash<T> const&>(_c)); }
 
-	using FixedHash<T>::operator==;
-	using FixedHash<T>::operator!=;
-	using FixedHash<T>::operator<;
-	using FixedHash<T>::operator>=;
-	using FixedHash<T>::operator<=;
-	using FixedHash<T>::operator>;
+    using FixedHash<T>::operator==;
+    using FixedHash<T>::operator!=;
+    using FixedHash<T>::operator<;
+    using FixedHash<T>::operator>=;
+    using FixedHash<T>::operator<=;
+    using FixedHash<T>::operator>;
 
-	// The obvious binary operators.
-	SecureFixedHash& operator^=(FixedHash<T> const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(_c); return *this; }
-	SecureFixedHash operator^(FixedHash<T> const& _c) const { return SecureFixedHash(*this) ^= _c; }
-	SecureFixedHash& operator|=(FixedHash<T> const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(_c); return *this; }
-	SecureFixedHash operator|(FixedHash<T> const& _c) const { return SecureFixedHash(*this) |= _c; }
-	SecureFixedHash& operator&=(FixedHash<T> const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(_c); return *this; }
-	SecureFixedHash operator&(FixedHash<T> const& _c) const { return SecureFixedHash(*this) &= _c; }
+    // The obvious binary operators.
+    SecureFixedHash& operator^=(FixedHash<T> const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(_c); return *this; }
+    SecureFixedHash operator^(FixedHash<T> const& _c) const { return SecureFixedHash(*this) ^= _c; }
+    SecureFixedHash& operator|=(FixedHash<T> const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(_c); return *this; }
+    SecureFixedHash operator|(FixedHash<T> const& _c) const { return SecureFixedHash(*this) |= _c; }
+    SecureFixedHash& operator&=(FixedHash<T> const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(_c); return *this; }
+    SecureFixedHash operator&(FixedHash<T> const& _c) const { return SecureFixedHash(*this) &= _c; }
 
-	SecureFixedHash& operator^=(SecureFixedHash const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(static_cast<FixedHash<T> const&>(_c)); return *this; }
-	SecureFixedHash operator^(SecureFixedHash const& _c) const { return SecureFixedHash(*this) ^= _c; }
-	SecureFixedHash& operator|=(SecureFixedHash const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(static_cast<FixedHash<T> const&>(_c)); return *this; }
-	SecureFixedHash operator|(SecureFixedHash const& _c) const { return SecureFixedHash(*this) |= _c; }
-	SecureFixedHash& operator&=(SecureFixedHash const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(static_cast<FixedHash<T> const&>(_c)); return *this; }
-	SecureFixedHash operator&(SecureFixedHash const& _c) const { return SecureFixedHash(*this) &= _c; }
-	SecureFixedHash operator~() const { auto r = ~static_cast<FixedHash<T> const&>(*this); return static_cast<SecureFixedHash const&>(r); }
+    SecureFixedHash& operator^=(SecureFixedHash const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(static_cast<FixedHash<T> const&>(_c)); return *this; }
+    SecureFixedHash operator^(SecureFixedHash const& _c) const { return SecureFixedHash(*this) ^= _c; }
+    SecureFixedHash& operator|=(SecureFixedHash const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(static_cast<FixedHash<T> const&>(_c)); return *this; }
+    SecureFixedHash operator|(SecureFixedHash const& _c) const { return SecureFixedHash(*this) |= _c; }
+    SecureFixedHash& operator&=(SecureFixedHash const& _c) { static_cast<FixedHash<T>&>(*this).operator^=(static_cast<FixedHash<T> const&>(_c)); return *this; }
+    SecureFixedHash operator&(SecureFixedHash const& _c) const { return SecureFixedHash(*this) &= _c; }
+    SecureFixedHash operator~() const { auto r = ~static_cast<FixedHash<T> const&>(*this); return static_cast<SecureFixedHash const&>(r); }
 
-	using FixedHash<T>::abridged;
-	using FixedHash<T>::abridgedMiddle;
+    using FixedHash<T>::abridged;
+    using FixedHash<T>::abridgedMiddle;
 
-	bytesConstRef ref() const { return FixedHash<T>::ref(); }
-	byte const* data() const { return FixedHash<T>::data(); }
+    bytesConstRef ref() const { return FixedHash<T>::ref(); }
+    byte const* data() const { return FixedHash<T>::data(); }
 
-	static SecureFixedHash<T> random() { SecureFixedHash<T> ret; ret.randomize(s_fixedHashEngine); return ret; }
-	using FixedHash<T>::firstBitSet;
+    static SecureFixedHash<T> random() { SecureFixedHash<T> ret; ret.randomize(s_fixedHashEngine); return ret; }
+    using FixedHash<T>::firstBitSet;
 
-	void clear() { ref().cleanse(); }
+    void clear() { ref().cleanse(); }
 };
 
 /// Fast equality operator for h256.
 template<> inline bool FixedHash<32>::operator==(FixedHash<32> const& _other) const
 {
-	const uint64_t* hash1 = (const uint64_t*)data();
-	const uint64_t* hash2 = (const uint64_t*)_other.data();
-	return (hash1[0] == hash2[0]) && (hash1[1] == hash2[1]) && (hash1[2] == hash2[2]) && (hash1[3] == hash2[3]);
+    const uint64_t* hash1 = (const uint64_t*)data();
+    const uint64_t* hash2 = (const uint64_t*)_other.data();
+    return (hash1[0] == hash2[0]) && (hash1[1] == hash2[1]) && (hash1[2] == hash2[2]) && (hash1[3] == hash2[3]);
 }
 
 /// Fast std::hash compatible hash function object for h256.
 template<> inline size_t FixedHash<32>::hash::operator()(FixedHash<32> const& value) const
 {
-	uint64_t const* data = reinterpret_cast<uint64_t const*>(value.data());
-	return boost::hash_range(data, data + 4);
+    uint64_t const* data = reinterpret_cast<uint64_t const*>(value.data());
+    return boost::hash_range(data, data + 4);
 }
 
 /// Stream I/O for the FixedHash class.
 template <unsigned N>
 inline std::ostream& operator<<(std::ostream& _out, FixedHash<N> const& _h)
 {
-	_out << toHex(_h);
-	return _out;
+    _out << toHex(_h);
+    return _out;
 }
 
 template <unsigned N>
 inline std::istream& operator>>(std::istream& _in, FixedHash<N>& o_h)
 {
-	std::string s;
-	_in >> s;
-	o_h = FixedHash<N>(s, FixedHash<N>::FromHex, FixedHash<N>::AlignRight);
-	return _in;
+    std::string s;
+    _in >> s;
+    o_h = FixedHash<N>(s, FixedHash<N>::FromHex, FixedHash<N>::AlignRight);
+    return _in;
 }
 
 /// Stream I/O for the SecureFixedHash class.
 template <unsigned N>
 inline std::ostream& operator<<(std::ostream& _out, SecureFixedHash<N> const& _h)
 {
-	_out << "SecureFixedHash#" << std::hex << typename FixedHash<N>::hash()(_h.makeInsecure()) << std::dec;
-	return _out;
+    _out << "SecureFixedHash#" << std::hex << typename FixedHash<N>::hash()(_h.makeInsecure()) << std::dec;
+    return _out;
 }
 
 // Common types of FixedHash.
@@ -349,9 +349,9 @@ using h160Hash = std::unordered_set<h160>;
 /// Convert the given value into h160 (160-bit unsigned integer) using the right 20 bytes.
 inline h160 right160(h256 const& _t)
 {
-	h160 ret;
-	memcpy(ret.data(), _t.data() + 12, 20);
-	return ret;
+    h160 ret;
+    memcpy(ret.data(), _t.data() + 12, 20);
+    return ret;
 }
 
 h128 fromUUID(std::string const& _uuid);
@@ -360,22 +360,22 @@ std::string toUUID(h128 const& _uuid);
 
 inline std::string toString(h256s const& _bs)
 {
-	std::ostringstream out;
-	out << "[ ";
-	for (h256 const& i: _bs)
-		out << i.abridged() << ", ";
-	out << "]";
-	return out.str();
+    std::ostringstream out;
+    out << "[ ";
+    for (h256 const& i: _bs)
+        out << i.abridged() << ", ";
+    out << "]";
+    return out.str();
 }
 
 }
 
 namespace std
 {
-	/// Forward std::hash<dev::FixedHash> to dev::FixedHash::hash.
-	template<> struct hash<dev::h64>: dev::h64::hash {};
-	template<> struct hash<dev::h128>: dev::h128::hash {};
-	template<> struct hash<dev::h160>: dev::h160::hash {};
-	template<> struct hash<dev::h256>: dev::h256::hash {};
-	template<> struct hash<dev::h512>: dev::h512::hash {};
+    /// Forward std::hash<dev::FixedHash> to dev::FixedHash::hash.
+    template<> struct hash<dev::h64>: dev::h64::hash {};
+    template<> struct hash<dev::h128>: dev::h128::hash {};
+    template<> struct hash<dev::h160>: dev::h160::hash {};
+    template<> struct hash<dev::h256>: dev::h256::hash {};
+    template<> struct hash<dev::h512>: dev::h512::hash {};
 }

--- a/libweb3jsonrpc/UnixSocketServer.cpp
+++ b/libweb3jsonrpc/UnixSocketServer.cpp
@@ -27,7 +27,7 @@ namespace fs = boost::filesystem;
 
 namespace
 {
-size_t const c_socketPathMaxLength = sizeof(sockaddr_un::sun_path) / sizeof(sockaddr_un::sun_path[0]);
+size_t const c_socketPathMaxLength = sizeof(sockaddr_un::sun_path) / sizeof(sockaddr_un::sun_path[0]) - 1;
 
 fs::path getIpcPathOrDataDir()
 {

